private/model/api/behavior.go

```diff
@@ -0,0 +1,320 @@
+// +build codegen
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+)
+
+type BehaviorTestSuite struct {
+	Defaults Defaults `json:"defaults"`
+	Tests    Tests    `json:"tests"`
+}
+
+type Tests struct {
+	Defaults Defaults `json:"defaults"`
+	Cases    []Case   `json:"cases"`
+}
+
+type Defaults struct {
+	Env    map[string]string `json:"env"`
+	Files  interface{}       `json:"files"`
+	Config map[string]string `json:"config"`
+}
+
+type Case struct {
+	Description string                   `json:"description"`
+	LocalEnv    map[string]string        `json:"env"`
+	LocalConfig map[string]string        `json:"config"`
+	Request     Request                  `json:"request"`
+	Response    Response                 `json:"response"`
+	Expect      []map[string]interface{} `json:"expect"`
+}
+
+type Response struct {
+	StatusCode  int               `json:"statusCode"`
+	BodyContent interface{}       `json:"bodyContent"`
+	BodyType    string            `json:"bodyType"`
+	Headers     map[string]string `json:"headers"`
+}
+
+type Request struct {
+	Operation string                 `json:"operation"`
+	Input     map[string]interface{} `json:"input"`
+}
+
+func (c Request) BuildInputShape(ref *ShapeRef) string {
+	b := NewShapeValueBuilder()
+	b.Base64BlobValues = true
+	return fmt.Sprintf("&%s{\n%s\n}",
+		b.GoType(ref, true),
+		b.BuildShape(ref, c.Input, false),
+	)
+}
+
+//Outputs the string to define an empty shape
+func (c Request) EmptyShapeBuilder(ref *ShapeRef) string {
+	b := NewShapeValueBuilder()
+	return fmt.Sprintf("%s{}", b.GoType(ref, true))
+}
+
+func (c Case) BuildOutputShape(ref *ShapeRef) string {
+	b := NewShapeValueBuilder()
+	b.Base64BlobValues = true
+	return fmt.Sprintf("&%s{\n%s\n}",
+		b.GoType(ref, true),
+		b.BuildShape(ref, c.Expect[0]["responseDataEquals"].(map[string]interface{}), false),
+	)
+}
+
+// AttachBehaviorTests attaches the Behavior test cases to the API model.
+func (a *API) AttachBehaviorTests(filename string) {
+	f, err := os.Open(filename)
+	if err != nil {
+		panic(fmt.Sprintf("failed to open behavior tests %s, err: %v", filename, err))
+	}
+	defer f.Close()
+
+	if err := json.NewDecoder(f).Decode(&a.BehaviorTests); err != nil {
+		panic(fmt.Sprintf("failed to decode behavior tests %s, err: %v", filename, err))
+	}
+
+}
+
+// APIBehaviorTestsGoCode returns the Go Code string for the Behavior tests.
+func (a *API) APIBehaviorTestsGoCode() string {
+	w := bytes.NewBuffer(nil)
```

Comment:
minor nit, this could be just `var w bytes.Buffer`

Reply:
`bytes.Buffer` does not implement `io.Writer` but the varible `w` uses `io.Writer` while executing the template